<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>gRNA_create.gRNA_genetic_algorithm API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>gRNA_create.gRNA_genetic_algorithm</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">from __future__ import annotations
from collections import Counter

from typing import Callable, List, cast, Dict

import pandas
import pygad
import seaborn

from gRNA_create.gRNA import gRNA
from gRNA_create.gRNA_scorer import Scorer
from gRNA_create.pam import PAM
from gRNA_create.utils import nucleotide_index_rna, ConfusionMatrix, \
    reverse_complement_dna
import numpy as np
from tqdm import tqdm
import matplotlib.pyplot as plt

gRNA_alias = List[int]


class gRNAGeneticAlgorithm(pygad.GA):

    def __init__(self,
                 num_generations: int,
                 num_parents_mating: int,
                 scoring_metric: Callable[[ConfusionMatrix], float],
                 initial_gRNAs: List[gRNA],
                 targets: List[str],
                 misses: List[str] = [],
                 mutation_rate: float = 0.1,
                 binding_cutoff: float = 0.5,
                 **kwargs
                 ):
        &#34;&#34;&#34;gRNA genetic algorithm for a signel gRNA location
        Args:
            num_generations (int): Number of generations to run algorithm for
            num_parents_mating (int): Number of gRNAs to splice together during &#34;mating&#34;
            scoring_metric (Callable[[ConfusionMatrix], float]): The scoring metric
            initial_gRNAs (List[gRNA]): The initial population of gRNAs
            targets (List[str]): The aligned target sequences
            misses (List[str], optional): The aligned negative sequences. Defaults to [].
            mutation_rate (float, optional): The chance of changing any nucleotide in any gRNA. Defaults to 0.1.
            binding_cutoff (float, optional): The cutoff score for binding. Defaults to 0.5.
        &#34;&#34;&#34;
        assert len(set([g.position for g in initial_gRNAs])) == 1
        assert len(set([g.pam for g in initial_gRNAs])) == 1
        assert len(set([g.scorer for g in initial_gRNAs])) == 1
        assert len(set([len(g) for g in initial_gRNAs])) == 1

        gRNA_aliases = [tuple([nucleotide_index_rna[nuc] for nuc in cur_gRNA]) for cur_gRNA in initial_gRNAs]

        self.pos = initial_gRNAs[0].position
        self.pam = initial_gRNAs[0].pam
        self.scorer = initial_gRNAs[0].scorer
        self.length = len(initial_gRNAs[0])

        self.scoring_metric = scoring_metric

        self.target_length = len(targets)
        self.miss_length = len(misses)

        self.binding_cutoff = binding_cutoff

        short_targets = [target[self.pos:self.pos + self.length] for target in targets]
        target_PAMs = [
            target[self.pos - self.pam.length: self.pos]
            if self.pam.location.value == 5
            else target[self.pos + self.length: self.pos + self.length + self.pam.length]
            for target in targets]
        self.PAM_and_targets_count = Counter(list(zip(target_PAMs, short_targets)))
        short_misses = [miss[self.pos:self.pos + self.length] for miss in misses]
        miss_PAMs = [miss[self.pos - self.pam.length: self.pos]
                     if self.pam.location.value == 5
                     else miss[self.pos + self.length: self.pos + self.length + self.pam.length]
                     for miss in misses] if misses is not None else []

        self.PAM_and_misses_count = Counter(list(zip(miss_PAMs, short_misses)))
        self.generation = 1

        def on_generation(x):
            if self.generation &gt; 1:
                end: str = &#34;&#34;
                if self.generation == num_generations:
                    end = &#34;\n&#34;
                print(&#34;\b&#34; * (len(str(self.generation - 1)) + 3 + len(str(num_generations))), end=end)
            else:
                print()
            print(self.generation, &#34;/&#34;, num_generations, end=&#34;&#34;)
            self.generation += 1

        super().__init__(
            num_generations=num_generations,
            num_parents_mating=num_parents_mating,
            gene_type=int,
            gene_space=range(4),
            initial_population=np.array(gRNA_aliases),
            fitness_func=_return_fitness_func(self.pos,
                                             self.pam,
                                             self.scorer,
                                             self.PAM_and_targets_count,
                                             self.PAM_and_misses_count,
                                             self.target_length,
                                             self.miss_length,
                                             self.scoring_metric,
                                             self.binding_cutoff),
            on_generation=on_generation,
            mutation_probability=mutation_rate,
            **kwargs
        )


def _return_fitness_func(pos: int, pam: PAM, scorer: Scorer, PAM_and_targets_count, PAM_and_misses_count, target_length,
                        miss_length, scoring_metric, binding_cutoff: float):
    def fitness_func(chromosome: gRNA_alias, idx: int):

        gRNA_str: str = &#34;&#34;.join([str(nucleotide_index_rna[nuc_index]) for nuc_index in chromosome])

        cur_gRNA: gRNA = gRNA(pos, gRNA_str, pam, scorer)

        PAM_and_targets = list(PAM_and_targets_count.keys())
        PAM_and_misses = list(PAM_and_misses_count.keys())

        target_PAMs, targets = zip(*PAM_and_targets)
        miss_PAMs, misses = zip(*PAM_and_misses) if miss_length &gt; 0 else ([], [])

        target_PAMs = [PAM(pam.location, p) for p in target_PAMs]
        miss_PAMs = [PAM(pam.location, p) for p in miss_PAMs]

        confusion_matrix: ConfusionMatrix = cast(ConfusionMatrix, {p: 0 for p in [&#34;tp&#34;, &#34;fn&#34;, &#34;tn&#34;, &#34;fp&#34;]})
        for i, res in enumerate(cur_gRNA.binds(targets, target_PAMs, cutoff=binding_cutoff)):
            confusion_matrix[&#34;tp&#34;] += res * PAM_and_targets_count[PAM_and_targets[i]]
        confusion_matrix[&#34;fn&#34;] = target_length - confusion_matrix[&#34;tp&#34;]
        if miss_length &gt; 0:
            for i, res in enumerate(cur_gRNA.binds(misses, miss_PAMs, cutoff=binding_cutoff)):
                confusion_matrix[&#34;fp&#34;] += res * PAM_and_misses_count[PAM_and_misses[i]]
            confusion_matrix[&#34;tn&#34;] = miss_length - confusion_matrix[&#34;fp&#34;]

        return scoring_metric(**confusion_matrix)

    return fitness_func


class gRNAGroupGA:

    def __init__(self,
                 num_generations: int,
                 num_parents_mating: int,
                 scoring_metric: Callable[[ConfusionMatrix], float],
                 initial_gRNAs: pandas.DataFrame,
                 targets: List[str],
                 misses: List[str] = [],
                 mutation_rate: float = 0.1,
                 multiplier: int = 1,
                 **kwargs
                 ):
        &#34;&#34;&#34;gRNA genetic algorithm for a whole population of gRNAs

        Args:
            num_generations (int): The number of generations to run the genetic algorithm for.
            num_parents_mating (int): The number of gRNAs to splice together.
            scoring_metric (Callable[[ConfusionMatrix], float]): The scoring metric to measure gRNA performance.
            initial_gRNAs (pandas.DataFrame): Initial population of gRNAs.
            targets (List[str]): List of aligned target regions.
            misses (List[str], optional): List of aligned negative regions. Defaults to [].
            mutation_rate (float, optional): The chance of inserting a mutation into any given position of any gRNA. Defaults to 0.1.
            multiplier (int, optional): The scalar to multiply the population of gRNAs by. Useful if there are not enough parents. Defaults to 1.
        &#34;&#34;&#34;
        self.pos_to_gRNAs: Dict[int, Dict[str, gRNA]] = {}
        self.scoring_metric: Callable[[ConfusionMatrix], float] = scoring_metric
        for _, cur_row in initial_gRNAs.iterrows():
            for _ in range(multiplier):
                if self.pos_to_gRNAs.get(cur_row[&#39;gRNA&#39;].position, -1) == -1:
                    self.pos_to_gRNAs[cur_row[&#39;gRNA&#39;].position] = cast(Dict[str, gRNA], {&#34;forward&#34;: [], &#34;reverse&#34;: []})
                if &#34;direction&#34; in cur_row and cur_row[&#34;direction&#34;] == &#34;reverse&#34;:
                    self.pos_to_gRNAs[cur_row[&#39;gRNA&#39;].position][&#34;reverse&#34;].append(cur_row[&#34;gRNA&#34;])
                else:
                    self.pos_to_gRNAs[cur_row[&#39;gRNA&#39;].position][&#34;forward&#34;].append(cur_row[&#34;gRNA&#34;])

        self.gRNA_genetic_algorithms: List[gRNAGeneticAlgorithm] = []
        for pos, g_dicts in self.pos_to_gRNAs.items():
            for dir, cur_gRNAs in g_dicts.items():
                if dir == &#34;reverse&#34; and len(cur_gRNAs) != 0:
                    self.gRNA_genetic_algorithms.append(
                        gRNAGeneticAlgorithm(num_generations, num_parents_mating, scoring_metric, cur_gRNAs,
                                             [reverse_complement_dna(cur_target) for cur_target in targets],
                                             [reverse_complement_dna(cur_miss) for cur_miss in misses],
                                             mutation_rate=mutation_rate, **kwargs))
                elif dir == &#34;forward&#34; and len(cur_gRNAs) != 0:
                    self.gRNA_genetic_algorithms.append(
                        gRNAGeneticAlgorithm(num_generations, num_parents_mating, scoring_metric, cur_gRNAs, targets,
                                             misses, mutation_rate=mutation_rate, **kwargs))

    @gRNA.parallelize()
    def run(self, show_end_graph: bool = False):
        &#34;&#34;&#34;Run the genetic algorithm

        Args:
            show_end_graph (bool, optional): Whether an end graph should be shown to see gRNA improvements. Defaults to False.
        &#34;&#34;&#34;
        df_pre = []

        for i, GA in tqdm(enumerate(self.gRNA_genetic_algorithms), total=len(self.gRNA_genetic_algorithms)):
            GA.run()
            df_pre += [{&#34;pos&#34;: list(self.pos_to_gRNAs.keys())[i], &#34;best&#34;: sol, &#34;generation&#34;: gen_num} for gen_num, sol
                       in enumerate(GA.best_solutions_fitness)]
        if show_end_graph:
            seaborn.lineplot(x=&#34;generation&#34;, y=&#34;best&#34;, data=pandas.DataFrame(df_pre), hue=&#34;pos&#34;)
            plt.show()

    def get_result(self) -&gt; pandas.DataFrame:
        &#34;&#34;&#34;Return genetic algorithm results

        Returns:
            pandas.DataFrame: Dataframe of new population of improved gRNAs.
        &#34;&#34;&#34;
        gRNA_to_metric_score: dict = {}
        for i, pos in enumerate(self.pos_to_gRNAs.keys()):
            cur_GA: gRNAGeneticAlgorithm = self.gRNA_genetic_algorithms[i]
            for chromosome, fitness in zip(cur_GA.population, cur_GA.last_generation_fitness):
                cur_gRNA: gRNA = gRNA(pos, &#34;&#34;.join([str(nucleotide_index_rna[nuc]) for nuc in chromosome]), cur_GA.pam,
                                      cur_GA.scorer)
                gRNA_to_metric_score[cur_gRNA] = fitness

        df_pre: List[dict] = \
            [{&#34;location&#34;: cur_gRNA.position, &#34;gRNA&#34;: cur_gRNA, self.scoring_metric.__name__: score} for cur_gRNA, score
             in gRNA_to_metric_score.items()]
        return pandas.DataFrame(df_pre)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="gRNA_create.gRNA_genetic_algorithm.gRNAGeneticAlgorithm"><code class="flex name class">
<span>class <span class="ident">gRNAGeneticAlgorithm</span></span>
<span>(</span><span>num_generations: int, num_parents_mating: int, scoring_metric: Callable[[ConfusionMatrix], float], initial_gRNAs: List[gRNA], targets: List[str], misses: List[str] = [], mutation_rate: float = 0.1, binding_cutoff: float = 0.5, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>gRNA genetic algorithm for a signel gRNA location</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>num_generations</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of generations to run algorithm for</dd>
<dt><strong><code>num_parents_mating</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of gRNAs to splice together during "mating"</dd>
<dt><strong><code>scoring_metric</code></strong> :&ensp;<code>Callable[[ConfusionMatrix], float]</code></dt>
<dd>The scoring metric</dd>
<dt><strong><code>initial_gRNAs</code></strong> :&ensp;<code>List[gRNA]</code></dt>
<dd>The initial population of gRNAs</dd>
<dt><strong><code>targets</code></strong> :&ensp;<code>List[str]</code></dt>
<dd>The aligned target sequences</dd>
<dt><strong><code>misses</code></strong> :&ensp;<code>List[str]</code>, optional</dt>
<dd>The aligned negative sequences. Defaults to [].</dd>
<dt><strong><code>mutation_rate</code></strong> :&ensp;<code>float</code>, optional</dt>
<dd>The chance of changing any nucleotide in any gRNA. Defaults to 0.1.</dd>
<dt><strong><code>binding_cutoff</code></strong> :&ensp;<code>float</code>, optional</dt>
<dd>The cutoff score for binding. Defaults to 0.5.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class gRNAGeneticAlgorithm(pygad.GA):

    def __init__(self,
                 num_generations: int,
                 num_parents_mating: int,
                 scoring_metric: Callable[[ConfusionMatrix], float],
                 initial_gRNAs: List[gRNA],
                 targets: List[str],
                 misses: List[str] = [],
                 mutation_rate: float = 0.1,
                 binding_cutoff: float = 0.5,
                 **kwargs
                 ):
        &#34;&#34;&#34;gRNA genetic algorithm for a signel gRNA location
        Args:
            num_generations (int): Number of generations to run algorithm for
            num_parents_mating (int): Number of gRNAs to splice together during &#34;mating&#34;
            scoring_metric (Callable[[ConfusionMatrix], float]): The scoring metric
            initial_gRNAs (List[gRNA]): The initial population of gRNAs
            targets (List[str]): The aligned target sequences
            misses (List[str], optional): The aligned negative sequences. Defaults to [].
            mutation_rate (float, optional): The chance of changing any nucleotide in any gRNA. Defaults to 0.1.
            binding_cutoff (float, optional): The cutoff score for binding. Defaults to 0.5.
        &#34;&#34;&#34;
        assert len(set([g.position for g in initial_gRNAs])) == 1
        assert len(set([g.pam for g in initial_gRNAs])) == 1
        assert len(set([g.scorer for g in initial_gRNAs])) == 1
        assert len(set([len(g) for g in initial_gRNAs])) == 1

        gRNA_aliases = [tuple([nucleotide_index_rna[nuc] for nuc in cur_gRNA]) for cur_gRNA in initial_gRNAs]

        self.pos = initial_gRNAs[0].position
        self.pam = initial_gRNAs[0].pam
        self.scorer = initial_gRNAs[0].scorer
        self.length = len(initial_gRNAs[0])

        self.scoring_metric = scoring_metric

        self.target_length = len(targets)
        self.miss_length = len(misses)

        self.binding_cutoff = binding_cutoff

        short_targets = [target[self.pos:self.pos + self.length] for target in targets]
        target_PAMs = [
            target[self.pos - self.pam.length: self.pos]
            if self.pam.location.value == 5
            else target[self.pos + self.length: self.pos + self.length + self.pam.length]
            for target in targets]
        self.PAM_and_targets_count = Counter(list(zip(target_PAMs, short_targets)))
        short_misses = [miss[self.pos:self.pos + self.length] for miss in misses]
        miss_PAMs = [miss[self.pos - self.pam.length: self.pos]
                     if self.pam.location.value == 5
                     else miss[self.pos + self.length: self.pos + self.length + self.pam.length]
                     for miss in misses] if misses is not None else []

        self.PAM_and_misses_count = Counter(list(zip(miss_PAMs, short_misses)))
        self.generation = 1

        def on_generation(x):
            if self.generation &gt; 1:
                end: str = &#34;&#34;
                if self.generation == num_generations:
                    end = &#34;\n&#34;
                print(&#34;\b&#34; * (len(str(self.generation - 1)) + 3 + len(str(num_generations))), end=end)
            else:
                print()
            print(self.generation, &#34;/&#34;, num_generations, end=&#34;&#34;)
            self.generation += 1

        super().__init__(
            num_generations=num_generations,
            num_parents_mating=num_parents_mating,
            gene_type=int,
            gene_space=range(4),
            initial_population=np.array(gRNA_aliases),
            fitness_func=_return_fitness_func(self.pos,
                                             self.pam,
                                             self.scorer,
                                             self.PAM_and_targets_count,
                                             self.PAM_and_misses_count,
                                             self.target_length,
                                             self.miss_length,
                                             self.scoring_metric,
                                             self.binding_cutoff),
            on_generation=on_generation,
            mutation_probability=mutation_rate,
            **kwargs
        )</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li>pygad.pygad.GA</li>
</ul>
</dd>
<dt id="gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA"><code class="flex name class">
<span>class <span class="ident">gRNAGroupGA</span></span>
<span>(</span><span>num_generations: int, num_parents_mating: int, scoring_metric: Callable[[ConfusionMatrix], float], initial_gRNAs: pandas.DataFrame, targets: List[str], misses: List[str] = [], mutation_rate: float = 0.1, multiplier: int = 1, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"><p>gRNA genetic algorithm for a whole population of gRNAs</p>
<h2 id="args">Args</h2>
<dl>
<dt><strong><code>num_generations</code></strong> :&ensp;<code>int</code></dt>
<dd>The number of generations to run the genetic algorithm for.</dd>
<dt><strong><code>num_parents_mating</code></strong> :&ensp;<code>int</code></dt>
<dd>The number of gRNAs to splice together.</dd>
<dt><strong><code>scoring_metric</code></strong> :&ensp;<code>Callable[[ConfusionMatrix], float]</code></dt>
<dd>The scoring metric to measure gRNA performance.</dd>
<dt><strong><code>initial_gRNAs</code></strong> :&ensp;<code>pandas.DataFrame</code></dt>
<dd>Initial population of gRNAs.</dd>
<dt><strong><code>targets</code></strong> :&ensp;<code>List[str]</code></dt>
<dd>List of aligned target regions.</dd>
<dt><strong><code>misses</code></strong> :&ensp;<code>List[str]</code>, optional</dt>
<dd>List of aligned negative regions. Defaults to [].</dd>
<dt><strong><code>mutation_rate</code></strong> :&ensp;<code>float</code>, optional</dt>
<dd>The chance of inserting a mutation into any given position of any gRNA. Defaults to 0.1.</dd>
<dt><strong><code>multiplier</code></strong> :&ensp;<code>int</code>, optional</dt>
<dd>The scalar to multiply the population of gRNAs by. Useful if there are not enough parents. Defaults to 1.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class gRNAGroupGA:

    def __init__(self,
                 num_generations: int,
                 num_parents_mating: int,
                 scoring_metric: Callable[[ConfusionMatrix], float],
                 initial_gRNAs: pandas.DataFrame,
                 targets: List[str],
                 misses: List[str] = [],
                 mutation_rate: float = 0.1,
                 multiplier: int = 1,
                 **kwargs
                 ):
        &#34;&#34;&#34;gRNA genetic algorithm for a whole population of gRNAs

        Args:
            num_generations (int): The number of generations to run the genetic algorithm for.
            num_parents_mating (int): The number of gRNAs to splice together.
            scoring_metric (Callable[[ConfusionMatrix], float]): The scoring metric to measure gRNA performance.
            initial_gRNAs (pandas.DataFrame): Initial population of gRNAs.
            targets (List[str]): List of aligned target regions.
            misses (List[str], optional): List of aligned negative regions. Defaults to [].
            mutation_rate (float, optional): The chance of inserting a mutation into any given position of any gRNA. Defaults to 0.1.
            multiplier (int, optional): The scalar to multiply the population of gRNAs by. Useful if there are not enough parents. Defaults to 1.
        &#34;&#34;&#34;
        self.pos_to_gRNAs: Dict[int, Dict[str, gRNA]] = {}
        self.scoring_metric: Callable[[ConfusionMatrix], float] = scoring_metric
        for _, cur_row in initial_gRNAs.iterrows():
            for _ in range(multiplier):
                if self.pos_to_gRNAs.get(cur_row[&#39;gRNA&#39;].position, -1) == -1:
                    self.pos_to_gRNAs[cur_row[&#39;gRNA&#39;].position] = cast(Dict[str, gRNA], {&#34;forward&#34;: [], &#34;reverse&#34;: []})
                if &#34;direction&#34; in cur_row and cur_row[&#34;direction&#34;] == &#34;reverse&#34;:
                    self.pos_to_gRNAs[cur_row[&#39;gRNA&#39;].position][&#34;reverse&#34;].append(cur_row[&#34;gRNA&#34;])
                else:
                    self.pos_to_gRNAs[cur_row[&#39;gRNA&#39;].position][&#34;forward&#34;].append(cur_row[&#34;gRNA&#34;])

        self.gRNA_genetic_algorithms: List[gRNAGeneticAlgorithm] = []
        for pos, g_dicts in self.pos_to_gRNAs.items():
            for dir, cur_gRNAs in g_dicts.items():
                if dir == &#34;reverse&#34; and len(cur_gRNAs) != 0:
                    self.gRNA_genetic_algorithms.append(
                        gRNAGeneticAlgorithm(num_generations, num_parents_mating, scoring_metric, cur_gRNAs,
                                             [reverse_complement_dna(cur_target) for cur_target in targets],
                                             [reverse_complement_dna(cur_miss) for cur_miss in misses],
                                             mutation_rate=mutation_rate, **kwargs))
                elif dir == &#34;forward&#34; and len(cur_gRNAs) != 0:
                    self.gRNA_genetic_algorithms.append(
                        gRNAGeneticAlgorithm(num_generations, num_parents_mating, scoring_metric, cur_gRNAs, targets,
                                             misses, mutation_rate=mutation_rate, **kwargs))

    @gRNA.parallelize()
    def run(self, show_end_graph: bool = False):
        &#34;&#34;&#34;Run the genetic algorithm

        Args:
            show_end_graph (bool, optional): Whether an end graph should be shown to see gRNA improvements. Defaults to False.
        &#34;&#34;&#34;
        df_pre = []

        for i, GA in tqdm(enumerate(self.gRNA_genetic_algorithms), total=len(self.gRNA_genetic_algorithms)):
            GA.run()
            df_pre += [{&#34;pos&#34;: list(self.pos_to_gRNAs.keys())[i], &#34;best&#34;: sol, &#34;generation&#34;: gen_num} for gen_num, sol
                       in enumerate(GA.best_solutions_fitness)]
        if show_end_graph:
            seaborn.lineplot(x=&#34;generation&#34;, y=&#34;best&#34;, data=pandas.DataFrame(df_pre), hue=&#34;pos&#34;)
            plt.show()

    def get_result(self) -&gt; pandas.DataFrame:
        &#34;&#34;&#34;Return genetic algorithm results

        Returns:
            pandas.DataFrame: Dataframe of new population of improved gRNAs.
        &#34;&#34;&#34;
        gRNA_to_metric_score: dict = {}
        for i, pos in enumerate(self.pos_to_gRNAs.keys()):
            cur_GA: gRNAGeneticAlgorithm = self.gRNA_genetic_algorithms[i]
            for chromosome, fitness in zip(cur_GA.population, cur_GA.last_generation_fitness):
                cur_gRNA: gRNA = gRNA(pos, &#34;&#34;.join([str(nucleotide_index_rna[nuc]) for nuc in chromosome]), cur_GA.pam,
                                      cur_GA.scorer)
                gRNA_to_metric_score[cur_gRNA] = fitness

        df_pre: List[dict] = \
            [{&#34;location&#34;: cur_gRNA.position, &#34;gRNA&#34;: cur_gRNA, self.scoring_metric.__name__: score} for cur_gRNA, score
             in gRNA_to_metric_score.items()]
        return pandas.DataFrame(df_pre)</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA.get_result"><code class="name flex">
<span>def <span class="ident">get_result</span></span>(<span>self) ‑> pandas.core.frame.DataFrame</span>
</code></dt>
<dd>
<div class="desc"><p>Return genetic algorithm results</p>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>pandas.DataFrame</code></dt>
<dd>Dataframe of new population of improved gRNAs.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_result(self) -&gt; pandas.DataFrame:
    &#34;&#34;&#34;Return genetic algorithm results

    Returns:
        pandas.DataFrame: Dataframe of new population of improved gRNAs.
    &#34;&#34;&#34;
    gRNA_to_metric_score: dict = {}
    for i, pos in enumerate(self.pos_to_gRNAs.keys()):
        cur_GA: gRNAGeneticAlgorithm = self.gRNA_genetic_algorithms[i]
        for chromosome, fitness in zip(cur_GA.population, cur_GA.last_generation_fitness):
            cur_gRNA: gRNA = gRNA(pos, &#34;&#34;.join([str(nucleotide_index_rna[nuc]) for nuc in chromosome]), cur_GA.pam,
                                  cur_GA.scorer)
            gRNA_to_metric_score[cur_gRNA] = fitness

    df_pre: List[dict] = \
        [{&#34;location&#34;: cur_gRNA.position, &#34;gRNA&#34;: cur_gRNA, self.scoring_metric.__name__: score} for cur_gRNA, score
         in gRNA_to_metric_score.items()]
    return pandas.DataFrame(df_pre)</code></pre>
</details>
</dd>
<dt id="gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA.run"><code class="name flex">
<span>def <span class="ident">run</span></span>(<span>*args, **kwargs)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def wrapper(*args, **kwargs):
    gRNA._pool = Pool(cores)
    try:
        res = method(*args, **kwargs)
        user_quit = False
    except BaseException:
        traceback.print_exc()
        user_quit = True
    gRNA._pool.close()
    gRNA._pool.join()

    if user_quit:
        sys.exit(1)
    return res</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="gRNA_create" href="index.html">gRNA_create</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="gRNA_create.gRNA_genetic_algorithm.gRNAGeneticAlgorithm" href="#gRNA_create.gRNA_genetic_algorithm.gRNAGeneticAlgorithm">gRNAGeneticAlgorithm</a></code></h4>
</li>
<li>
<h4><code><a title="gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA" href="#gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA">gRNAGroupGA</a></code></h4>
<ul class="">
<li><code><a title="gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA.get_result" href="#gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA.get_result">get_result</a></code></li>
<li><code><a title="gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA.run" href="#gRNA_create.gRNA_genetic_algorithm.gRNAGroupGA.run">run</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>